snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,340 @@
+            # taking the lock):
+            if message.value.consumer_group in self.__remote_consumer_groups:
+                with self.__subscription.get() as subscription:
+                    # If the current subscription has changed while we were
+                    # waiting for a message, we need to discard this message.
+                    # The next loop iteration will update the consumer
+                    # assignment, and we will begin to rebuild state there.
+                    if subscription is not current_subscription:
+                        continue
+
+                    # If the stream for this message matches the subscription
+                    # topics, apply the update to the subscription offsets.
+                    if any((message.stream in topic) for topic in subscription.topics):
+                        subscription.remote_consumer_group_offsets[
+                            message.value.stream
+                        ][message.value.consumer_group] = message.value.offset
+
+    def subscribe(
+        self,
+        topics: Sequence[
+            Topic[TStream]
+        ],  # TODO: This will have to apply to the superclass as well.
+        on_assign: Optional[Callable[[Mapping[TStream, TOffset]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[TStream]], None]] = None,
+    ) -> None:
+        # Update the subscription, so that the commit log consumer can starts
+        # to consume the commit log data for the subscribed (and potentially
+        # assigned) topics.
+        self.__subscription.set(Subscription(topics))
+
+        # TODO: How does the Kafka consumer handle paused streams? My
+        # assumption here is that the new subscription should cause those
+        # streams to be resumed again, but not sure that's accurate.
+        self.__paused_streams.clear()
+
+        def assignment_callback(streams: Mapping[TStream, TOffset]) -> None:
+            with self.__subscription.get() as subscription:
+                for stream, offset in streams.items():
+                    remote_offset = subscription.get_effective_remote_consumer_group_offset(
+                        stream, self.__remote_consumer_groups
+                    )
+                    if remote_offset is None or offset >= remote_offset:
+                        self.__backend.pause([stream])
+                    elif stream not in self.__paused_streams:
+                        # TODO: It should be safe to call resume on a stream
+                        # that is already consuming, but may have some
+                        # performance drawbacks (I vaguely recall that this may
+                        # flush the consume buffer in the Confluent consumer),
+                        # so it may be worth caching which streams are active
+                        # or paused here, or in the KafkaConsumerBackend.
+                        self.__backend.resume([stream])
```

Comment:
Parts of this block are duplicated in other places (in here, `poll` twice, `seek`, and `resume`) and could probably be consolidated somewhat.

Reply:
Revised this with 0c1b32ce5a66e9dbab0e52df219986d2c6aa6523 for the common cases. The second remote offset check in `poll` is still a bit of a special case, since we need to know whether that should pause *and* trigger an offset rollback.